integrationtests/test_all_examples.py

```diff
@@ -0,0 +1,281 @@
+#  Copyright (c) 2020, Enthought, Inc.
+#  All rights reserved.
+#
+#  This software is provided without warranty under the terms of the BSD
+#  license included in LICENSE.txt and may be redistributed only
+#  under the conditions described in the aforementioned license.  The license
+#  is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+#  Thanks for using Enthought open source!
+
+""" Tests for demo and tutorial examples.
+"""
+
+import contextlib
+import io
+from itertools import chain
+import os
+import sys
+import traceback
+import unittest
+from unittest import mock
+
+from pyface.api import GUI
+from traits.api import HasTraits
+
+from traitsui.tests._tools import (
+    is_current_backend_wx,
+    is_current_backend_qt4,
+    skip_if_null,
+    store_exceptions_on_all_threads,
+)
+
+
+# This test file is not distributed nor is it in a package.
+HERE = os.path.dirname(__file__)
+
+EXAMPLES = os.path.join(HERE, "..", "examples")
+DEMO = os.path.join(EXAMPLES, "demo")
+TUTORIALS = os.path.join(EXAMPLES, "tutorials", "doc_examples", "examples")
+
+#: Explicitly include folders from which example files should be found
+#: recursively.
+SOURCE_DIRS = [
+    DEMO,
+    TUTORIALS,
+]
+
+
+#: Mapping from filepath to a callable() -> (skipped: bool, reason: str)
+FILES_MAY_BE_SKIPPED = {}
+
+
+def skip_file_if(condition, reason, filepath):
```

Comment:
Maybe I'm missing some intended capabilities, but I find this `skip_file_if` system too complex for what it does. Why not have a simple dictionary with structured entries, that I think would be easier to read, e.g.:
``` python
FILES_MAY_BE_SKIPPED = {
    os.path.join(DEMO, "demo.py"): ( 
        lambda: True,
        "Not a target file to be tested"
    )
}
```
`os.path.abspath` shouldn't be necessary as long as `EXAMPLES` is saved as an absolute path. And I don't think condition execution is needed at import time, it could be done within the subtest by modifying `is_skipped`:
```python
    skip_condition, reason = FILES_MAY_BE_SKIPPED[path]
    return skip_condition(), reason
```

Reply:
Good point. Not sure why I did that. The intermediate wrapper is uncalled for.

Reply:
The `os.path.abspath` is important actually. It removes the assumption that file path discovery must match the configuration for what files should be skipped.

Reply:
I'm not sure I follow that. The assumption that the keys of `FILES_MAY_BE_SKIPPED` will match the paths collected during file discovery is there even with `os.path.abspath`. It's the assumption that both will be saved as absolute paths.